# LLM warning:
#
# This patch was generated by Jan "Yenya" Kasprzak using
# Kimi 3 large language model. The prompt was:
#
# The stepmania source rpm in this directory uses a deprecated
# pcre version 1 library. write a patch file which provides support
# for using the system pcre2 library instead, integrate it to the spec file,
# and verify that the package can be correctly compiled. Show also the diff
# to the original .spec file.
# The original src.rpm should build on this machine without needing to install
# other dependencies, everything is already installed. PCRE2 docs are also
# installed. Don't run sudo for installing anything else.
#
#
diff -uN -r stepmania/CMake/DefineOptions.cmake stepmania/CMake/DefineOptions.cmake
--- stepmania/CMake/DefineOptions.cmake	2026-08-26 18:06:51.818682810 +0200
+++ stepmania/CMake/DefineOptions.cmake	2026-08-26 18:07:16.072977384 +0200
@@ -60,7 +60,7 @@
 option(WITH_SYSTEM_MAD "Build with system libmad" OFF)
 option(WITH_SYSTEM_JSONCPP "Build with system jsoncpp" OFF)
 option(WITH_SYSTEM_JPEG "Build with system jpeglib" OFF)
-option(WITH_SYSTEM_PCRE "Build with system PCRE" OFF)
+option(WITH_SYSTEM_PCRE "Build with system PCRE2" OFF)
 option(WITH_SYSTEM_ZLIB "Build against system zlib" OFF)
 
 option(WITH_SDL "Build with SDL" OFF)
diff -uN -r stepmania/CMake/Modules/FindPCRE2.cmake stepmania/CMake/Modules/FindPCRE2.cmake
--- stepmania/CMake/Modules/FindPCRE2.cmake	1970-01-01 01:00:00.000000000 +0100
+++ stepmania/CMake/Modules/FindPCRE2.cmake	2026-08-26 18:16:59.442443995 +0200
@@ -0,0 +1,17 @@
+# Find pcre2 using standard tools.
+
+# The following will be set:
+
+# PCRE2_INCLUDE_DIR PCRE2_LIBRARY PCRE2_FOUND
+
+find_path(PCRE2_INCLUDE_DIR NAMES pcre2.h)
+find_library(PCRE2_LIBRARY NAMES pcre2-8)
+
+# Properly pass QUIETLY and REQUIRED.
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(PCRE2
+                                  DEFAULT_MSG
+                                  PCRE2_INCLUDE_DIR
+                                  PCRE2_LIBRARY)
+
+mark_as_advanced(PCRE2_INCLUDE_DIR PCRE2_LIBRARY)
diff -uN -r stepmania/CMake/Modules/FindPcre.cmake stepmania/CMake/Modules/FindPcre.cmake
--- stepmania/CMake/Modules/FindPcre.cmake	2026-08-26 18:06:51.818983353 +0200
+++ stepmania/CMake/Modules/FindPcre.cmake	1970-01-01 01:00:00.000000000 +0100
@@ -1,17 +0,0 @@
-# Find pcre using standard tools.
-
-# The following will be set:
-
-# PCRE_INCLUDE_DIR PCRE_LIBRARY PCRE_FOUND
-
-find_path(PCRE_INCLUDE_DIR NAMES pcre.h)
-find_library(PCRE_LIBRARY NAMES pcre)
-
-# Properly pass QUIETLY and REQUIRED.
-include(FindPackageHandleStandardArgs)
-find_package_handle_standard_args(PCRE
-                                  DEFAULT_MSG
-                                  PCRE_INCLUDE_DIR
-                                  PCRE_LIBRARY)
-
-mark_as_advanced(PCRE_INCLUDE_DIR PCRE_LIBRARY)
diff -uN -r stepmania/extern/CMakeProject-pcre.cmake stepmania/extern/CMakeProject-pcre.cmake
--- stepmania/extern/CMakeProject-pcre.cmake	2026-08-26 18:06:52.047678931 +0200
+++ stepmania/extern/CMakeProject-pcre.cmake	2026-08-26 18:16:48.424301862 +0200
@@ -1,5 +1,5 @@
 if(WITH_SYSTEM_PCRE)
-  find_package(Pcre REQUIRED)
+  find_package(PCRE2 REQUIRED)
 else()
   set(PCRE_SRC "pcre/get.c" "pcre/maketables.c" "pcre/pcre.c" "pcre/study.c")
 
diff -uN -r stepmania/src/CMakeLists.txt stepmania/src/CMakeLists.txt
--- stepmania/src/CMakeLists.txt	2026-08-26 18:06:52.160668912 +0200
+++ stepmania/src/CMakeLists.txt	2026-08-26 18:23:22.501428319 +0200
@@ -438,7 +438,9 @@
 endif()
 
 if(WITH_SYSTEM_PCRE)
-  list(APPEND SMDATA_LINK_LIB ${PCRE_LIBRARY})
+  list(APPEND SMDATA_LINK_LIB ${PCRE2_LIBRARY})
+  sm_add_compile_definition("${SM_EXE_NAME}" USE_SYSTEM_PCRE)
+  sm_add_compile_definition("${SM_EXE_NAME}" PCRE2_CODE_UNIT_WIDTH=8)
 else()
   list(APPEND SMDATA_LINK_LIB "pcre")
 endif()
@@ -631,10 +633,10 @@
     sm_add_compile_definition("${SM_EXE_NAME}" HAVE_SDL)
   endif()
 
-  if(PCRE_FOUND)
-    list(APPEND SMDATA_LINK_LIB ${PCRE_LIBRARY})
+  if(PCRE2_FOUND)
+    list(APPEND SMDATA_LINK_LIB ${PCRE2_LIBRARY})
   endif()
-  
+
   if(LIBXTST_FOUND)
     list(APPEND SMDATA_LINK_LIB ${LIBXTST_LIBRARY})
   endif()
@@ -693,8 +695,8 @@
     if(X11_FOUND)
       list(APPEND SM_INCLUDE_DIRS "${X11_INCLUDE_DIR}")
     endif()
-    if(PCRE_FOUND)
-      list(APPEND SM_INCLUDE_DIRS "${PCRE_INCLUDE_DIR}")
+    if(PCRE2_FOUND)
+      list(APPEND SM_INCLUDE_DIRS "${PCRE2_INCLUDE_DIR}")
     endif()
     if(DL_FOUND)
       list(APPEND SM_INCLUDE_DIRS "${DL_INCLUDE_DIR}")
diff -uN -r stepmania/src/RageUtil.cpp stepmania/src/RageUtil.cpp
--- stepmania/src/RageUtil.cpp	2026-08-26 18:06:52.184669222 +0200
+++ stepmania/src/RageUtil.cpp	2026-08-26 18:11:15.633065518 +0200
@@ -1467,18 +1467,24 @@
 #ifndef USE_SYSTEM_PCRE
 #include "../extern/pcre/pcre.h"
 #else
-#include <pcre.h>
+#include <pcre2.h>
 #endif
 void Regex::Compile()
 {
-	const char *error;
-	int offset;
-	m_pReg = pcre_compile( m_sPattern.c_str(), PCRE_CASELESS, &error, &offset, nullptr );
+	int error;
+	PCRE2_SIZE offset;
+	m_pReg = (void *) pcre2_compile( (PCRE2_SPTR) m_sPattern.c_str(), PCRE2_ZERO_TERMINATED,
+		PCRE2_CASELESS, &error, &offset, nullptr );
 
 	if( m_pReg == nullptr )
-		RageException::Throw( "Invalid regex: \"%s\" (%s).", m_sPattern.c_str(), error );
+	{
+		PCRE2_UCHAR sError[256];
+		pcre2_get_error_message( error, sError, sizeof(sError) );
+		RageException::Throw( "Invalid regex: \"%s\" (%s).", m_sPattern.c_str(), (const char *) sError );
+	}
 
-	int iRet = pcre_fullinfo( (pcre *) m_pReg, nullptr, PCRE_INFO_CAPTURECOUNT, &m_iBackrefs );
+	m_pMatch = (void *) pcre2_match_data_create_from_pattern( (pcre2_code *) m_pReg, nullptr );
+	int iRet = pcre2_pattern_info( (pcre2_code *) m_pReg, PCRE2_INFO_CAPTURECOUNT, &m_iBackrefs );
 	ASSERT( iRet >= 0 );
 
 	++m_iBackrefs;
@@ -1494,17 +1500,22 @@
 
 void Regex::Release()
 {
-	pcre_free( m_pReg );
+	if( m_pMatch != nullptr )
+	{
+		pcre2_match_data_free( (pcre2_match_data *) m_pMatch );
+		m_pMatch = nullptr;
+	}
+	pcre2_code_free( (pcre2_code *) m_pReg );
 	m_pReg = nullptr;
 	m_sPattern = RString();
 }
 
-Regex::Regex( const RString &sStr ): m_pReg(nullptr), m_iBackrefs(0), m_sPattern(RString())
+Regex::Regex( const RString &sStr ): m_pReg(nullptr), m_pMatch(nullptr), m_iBackrefs(0), m_sPattern(RString())
 {
 	Set( sStr );
 }
 
-Regex::Regex( const Regex &rhs ): m_pReg(nullptr), m_iBackrefs(0), m_sPattern(RString())
+Regex::Regex( const Regex &rhs ): m_pReg(nullptr), m_pMatch(nullptr), m_iBackrefs(0), m_sPattern(RString())
 {
 	Set( rhs.m_sPattern );
 }
@@ -1523,11 +1534,11 @@
 
 bool Regex::Compare( const RString &sStr )
 {
-	int iMat[128*3];
-	int iRet = pcre_exec( (pcre *) m_pReg, nullptr, sStr.data(), sStr.size(), 0, 0, iMat, 128*3 );
+	int iRet = pcre2_match( (pcre2_code *) m_pReg, (PCRE2_SPTR) sStr.data(), sStr.size(),
+		0, 0, (pcre2_match_data *) m_pMatch, nullptr );
 
 	if( iRet < -1 )
-		RageException::Throw( "Unexpected return from pcre_exec('%s'): %i.", m_sPattern.c_str(), iRet );
+		RageException::Throw( "Unexpected return from pcre2_match('%s'): %i.", m_sPattern.c_str(), iRet );
 
 	return iRet >= 0;
 }
@@ -1536,15 +1547,17 @@
 {
 	asMatches.clear();
 
-	int iMat[128*3];
-	int iRet = pcre_exec( (pcre *) m_pReg, nullptr, sStr.data(), sStr.size(), 0, 0, iMat, 128*3 );
+	int iRet = pcre2_match( (pcre2_code *) m_pReg, (PCRE2_SPTR) sStr.data(), sStr.size(),
+		0, 0, (pcre2_match_data *) m_pMatch, nullptr );
 
 	if( iRet < -1 )
-		RageException::Throw( "Unexpected return from pcre_exec('%s'): %i.", m_sPattern.c_str(), iRet );
+		RageException::Throw( "Unexpected return from pcre2_match('%s'): %i.", m_sPattern.c_str(), iRet );
 
 	if( iRet == -1 )
 		return false;
 
+	PCRE2_SIZE *iMat = pcre2_get_ovector_pointer( (pcre2_match_data *) m_pMatch );
+
 	for( unsigned i = 1; i < m_iBackrefs; ++i )
 	{
 		const int iStart = iMat[i*2], end = iMat[i*2+1];
diff -uN -r stepmania/src/RageUtil.h stepmania/src/RageUtil.h
--- stepmania/src/RageUtil.h	2026-08-26 18:06:52.184669222 +0200
+++ stepmania/src/RageUtil.h	2026-08-26 18:10:57.998838499 +0200
@@ -552,6 +552,7 @@
 	void Release();
 
 	void *m_pReg;
+	void *m_pMatch;
 	unsigned m_iBackrefs;
 	RString m_sPattern;
 };
